bam_addrprg.c

```diff
@@ -0,0 +1,471 @@
+
+    retval->output_header = bam_hdr_dup(retval->input_header);
+    retval->output_file = sam_open_format(opts->output_name, opts->output_mode, &opts->ga.out);
+    
+    if (retval->output_file == NULL) {
+        print_error_errno("Could not open output file: %s mode: %s\n", opts->output_name, opts->output_mode);
+        return true;
+    }
+
+    if (opts->rg_line) {
+        // Append new RG line to header.
+        // Check does not already exist
+        if ( confirm_rg(retval->output_header, opts->rg_id) ) {
+            fprintf(stderr, "[init] ID of new RG line specified conflicts with that of an existing header RG line. Overwrite not yet implemented.\n");
+            return true;
+        }
+        retval->rg_id = strdup(opts->rg_id);
+        size_t new_len = strlen( retval->output_header->text ) + strlen( opts->rg_line ) + 2;
+        char* new_header = malloc(new_len);
+        if (!new_header) {
+            fprintf(stderr, "[init] Out of memory whilst writing new header.\n");
+            return true;
+        }
+        sprintf(new_header,"%s%s\n", retval->output_header->text, opts->rg_line);
+        free(retval->output_header->text);
+        retval->output_header->text = new_header;
+        retval->output_header->l_text = (int)new_len;
+    } else {
+        if (opts->rg_id) {
+            // Confirm what has been supplied exists
+            if ( !confirm_rg(retval->output_header, opts->rg_id) ) {
+                fprintf(stderr, "RG ID supplied does not exist in header. Supply full @RG line with -r instead?\n");
+                return true;
+            }
+            retval->rg_id = strdup(opts->rg_id);
+        } else {
+            if ((retval->rg_id = get_first_rgid(retval->output_header)) == NULL ) {
+                fprintf(stderr, "No RG specified on command line or in existing header.\n");
+                return true;
+            }
+        }
+    }
+
+    switch (opts->mode) {
+        case overwrite_all:
+            retval->mode_func = &overwrite_all_func;
+            break;
+        case orphan_only:
+            retval->mode_func = &orphan_only_func;
+            break;
+    }
+
+    return false;
+}
+
+static bool readgroupise(state_t* state)
+{
+    if (sam_hdr_write(state->output_file, state->output_header) != 0) {
+        return true;
+    }
+
+    bam1_t* file_read = bam_init1();
+    if (sam_read1(state->input_file, state->input_header, file_read) < 0) {
+        bam_destroy1(file_read);
+        file_read = NULL;
+    }
+    while (file_read != NULL) {
+        state->mode_func(state, file_read);
+
+        if (sam_write1(state->output_file, state->output_header, file_read) < 0) {
+            fprintf(stderr, "Could not write read to output file.\n");
+            return true;
+        }
+        if (sam_read1(state->input_file, state->input_header, file_read) < 0) {
+            bam_destroy1(file_read);
+            file_read = NULL;
+        }
+    }
```

Comment:
Is there a reason why this loop is not written in the usual `while (sam_read1(...) >= 0)` way?